crates/m3/tests/merkle_tree.rs

```diff
@@ -0,0 +1,401 @@
+	type NodeFlush = (u8, [u8; 32], u32, u32);
+	// Signature of the Roots channel: (Root ID, Root digest)
+	type RootFlush = (u8, [u8; 32]);
+
+	/// A type alias for the Merkle path, which is a vector of tuples containing the root ID, index,
+	/// leaf, and the path to the leaf.
+	type MerklePath<'a> = &'a [(u8, u32, [u8; 32], Vec<[u8; 32]>)];
+
+	/// A struct whose fields contain the channels involved in the trace to verify merkle paths for
+	/// a binary merkle tree
+	pub struct MerkleTreeChannels {
+		/// This channel gets flushed with tokens during "intermediate" steps of the verification
+		/// where the tokens are the values of the parent digest along with associated position
+		/// information such as the root it is associated to, the value of the digest, the depth
+		/// and the index.
+		nodes: Channel<NodeFlush>,
+
+		/// This channel contains flushes that validate that the "final" digest obtained in a
+		/// merkle path is matches that of one of the claimed roots, pushed as boundary values.
+		roots: Channel<RootFlush>,
+	}
+
+	impl MerkleTreeChannels {
+		pub fn new() -> Self {
+			Self {
+				nodes: Channel::default(),
+				roots: Channel::default(),
+			}
+		}
+	}
+	/// A token representing a step in verifying a merkle path for inclusion.
+	pub struct MerklePathToken {
+		pub root_id: u8,
+		pub left: [u8; 32],
+		pub right: [u8; 32],
+		pub parent: [u8; 32],
+		pub parent_depth: u32,
+		pub parent_index: u32,
+		pub flush_left: bool,
+		pub flush_right: bool,
+	}
+
+	/// A token representing the final step of comparing the claimed root.
+	pub struct MerkleRootToken {
+		pub root_id: u8,
+		pub digest: [u8; 32],
+	}
+	#[allow(clippy::too_many_arguments)]
+	impl MerklePathToken {
+		pub fn new(
+			root_id: u8,
+			left: [u8; 32],
+			right: [u8; 32],
+			parent: [u8; 32],
+			parent_depth: u32,
+			parent_index: u32,
+			flush_left: bool,
+			flush_right: bool,
+		) -> Self {
+			Self {
+				root_id,
+				left,
+				right,
+				parent,
+				parent_depth,
+				parent_index,
+				flush_left,
+				flush_right,
+			}
+		}
+	}
+
+	impl MerkleRootToken {
+		pub fn new(root_id: u8, digest: [u8; 32]) -> Self {
+			Self { root_id, digest }
+		}
+	}
+
+	// Uses the Groestl256 compression function to compress two 32-byte inputs into a single 32-byte
+	fn compress(left: &[u8], right: &[u8], output: &mut [u8]) {
+		let mut state_bytes = [0u8; 64];
+		let (half0, half1) = state_bytes.split_at_mut(32);
+		half0.copy_from_slice(left);
+		half1.copy_from_slice(right);
+		let input = GroestlShortImpl::state_from_bytes(&state_bytes);
+		let mut state = input;
+		GroestlShortImpl::p_perm(&mut state);
+		GroestlShortImpl::xor_state(&mut state, &input);
+		state_bytes = GroestlShortImpl::state_to_bytes(&state);
+		output.copy_from_slice(&state_bytes[32..]);
+	}
+
+	// Merkle tree implementation for the model, assumes the leaf layer consists of [u8;32] blobs.
+	pub struct MerkleTree {
+		depth: usize,
+		nodes: Vec<[u8; 32]>,
```

Comment:
tiny nit: maybe worth leaving a comment about the lay out of the nodes.

Feel free to ignore this one.